### ISSUE TITLE
Drake versioned releases do not self-identify

### ISSUE THREAD

**Issue by stonier** [RobotLocomotion/drake#14509]
Or at least, I haven't found anything in a grep of a versioned release (e.g. [drake-0.25.0-bionic.tar.gz](https://drake-packages.csail.mit.edu/drake/release/drake-0.25.0-bionic.tar.gz). The drake `VERSION.txt` file provides only the date / commit hash.

What would be useful:

* Drake's `VERSION.txt` can communicate version possibly with the ability to distinguish between snapshot (nightly) and release binaries.
* C++ version macros in headers
* Python version numbers from modules a'la [PEP-0396](https://www.python.org/dev/peps/pep-0396/)

The first one at least would be a good starting point - that will allow humans and scripts to introspect what's installed more efficiently than having to associate with releases on the server.

**Comment by jamiesnape** [comment]
Yes, this is an issue for me too. The version number is effectively decided after the event (the release notes are missing from the commit that forms the release). It will need fixing up by tooling or manually, though changing the VERSION.txt changes the SHA, so really we need to fix this before the packages go to GitHub, rather than before they go to S3.

I have to fix this for apt packaging too.

**Comment by jamiesnape** [comment]
Probably having Jenkins have a job that takes a SHA parameter and a release number would be a start. We are also missing `drake-config-version.cmake` for CMake.

**Comment by jamiesnape** [comment]
Also Bazel has the ability to embed release information, which we do not use at present.

**Comment by jamiesnape** [comment]
(To me this is high priority, that is, not necessarily that it should be labeled high priority, but it would make my life disproportionately easier if we achieved this, especially if the release notes were in the release.)

**Comment by stonier** [comment]
> We are also missing drake-config-version.cmake for CMake.

Just went looking for that this morning (needed it). Added it to the list in the issue description.

With regards to priorities, +1 from me. Justifying it's priority really ought to be derived from the priority allocated to user support - this one is a fundamental release mechanism for users. It's rarely used internally by developers (@jamiesnape I imagine you're the exception), so justifying it's priority from that angle will always be difficult.

**Comment by jamiesnape** [comment]
> (@jamiesnape I imagine you're the exception),

I am always an exception

But seriously, I need version numbers and release notes (or just release notes that have a clear regex for version number) so that they make to the users in the multiple places they need embedding and I am a proto-user who is supposed to show people how to use the version numbers in the examples. I actually need the tagged commit to contain at least the code to generate the correctly versioned files.

**Comment by jamiesnape** [comment]
> C++ version macros in headers

This could be achieved via some Bazel and  C++ code and the flag `--embed_label` IIRC.

**Comment by jwnimmer-tri** [comment]
The list of places to note a version badge in the top post is sound, and a great idea.

On the topic of priority, this is one of the most important issues in its subject area, but we just have a lot of projects already in flight (apt packages, catching up on macOS support, website refresh) that we need to carry over the finish line before revisiting the versioned-release machinery.

**Comment by jwnimmer-tri** [comment]
Update on the status quo:

For our binary packages ([from_binary](https://drake.mit.edu/from_binary.html) and [apt](https://drake.mit.edu/apt.html)), the text file `drake/share/doc/drake/VERSION.txt` contains the version number and git sha (`1.51.1 778ac9e2147ae4b9053db8489196f9d38cd5d9ad`).

The `drake-config.cmake` also provides the version number:

```cmake
set(${CMAKE_FIND_PACKAGE_NAME}_VERSION "1.51.1")
set(${CMAKE_FIND_PACKAGE_NAME}_VERSION_MAJOR "1")
set(${CMAKE_FIND_PACKAGE_NAME}_VERSION_MINOR "51")
set(${CMAKE_FIND_PACKAGE_NAME}_VERSION_PATCH "1")
set(${CMAKE_FIND_PACKAGE_NAME}_VERSION_TWEAK "0")
```

For our wheel packages ([pip](https://drake.mit.edu/pip.html)), the text file `pydrake/doc/drake/VERSION.TXT` contains the same (`1.51.1 778a...`), but users will probably prefer the canonical `importlib` mechanism:

```console
jwnimmer@call-cps:~/tmp$ python3 -m venv env
jwnimmer@call-cps:~/tmp$ env/bin/pip install drake
...
jwnimmer@call-cps:~/tmp$ env/bin/python3
>>> import importlib.metadata
>>> importlib.metadata.version("drake")
'1.51.1'
```

Nightly snapshots have a version number like `0.0.20260403` instead of `1.51.1`.  Experimental snapshots have a version like `0.0.20260403.130226+gite6225368`.

That leaves the following requests open:

> - Python version numbers from modules a'la [PEP-0396](https://www.python.org/dev/peps/pep-0396/).
> - `drake.__version__` in pydrake.

That PEP was withdrawn.  The importlib spelling is preferred now, and already works.  Similarly, `__version__` is dis-preferred in favor of importlib.  We don't plan on implementing either one.

> - Bazel versioning in the release.

As shown with [drake_bazel_external](https://github.com/RobotLocomotion/drake-external-examples/tree/main/drake_bazel_external), the version number isn't required for downstream users with Bazel.  (There isn't a way to programmatically access it, anyway.)  This would become relevant if and only if Drake was published in a bzlmod registry somewhere.  BCR publication is tracked in #23301, so we'll defer to that issue on any questions of Bazel-specific version stamping in a release.

> - C++ version macros in headers.

This is not yet implemented.

**Comment by jwnimmer-tri** [comment]
#24342 will change the python version self-reporting to be uniform across all of our release channels, for additional convenience.

The only remaining work on this ticket is to provide C++ preprocessor macros.  I've filed #24343 to track that as a more specific request, and we'll call this meta-issue completed.

**Comment by tyler-yankee** [comment]
(I don't generally edit other folks' comments, but I've edited https://github.com/RobotLocomotion/drake/issues/14509#issuecomment-4184987772 to correct the link to the new C++ version macro issue.)
